engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -147,52 +158,50 @@ protected void beginTransaction(@NotNull final Object token) {
      */
     protected void endTransaction(@NotNull final Object token) {
         final Transaction transaction;
-        synchronized (transactionLock) {
+        synchronized (transactions) {
             // Verify that this transaction is open.
             transaction = transactions.remove(token);
             if (transaction == null) {
                 throw new IllegalStateException("No transaction with token " + token + " is currently open.");
             }
         }
 
-        final Collection<ImmutableTableLocationKey> addedImmutableKeys =
+        final Collection<ImmutableTableLocationKey> addedKeys =
                 new ArrayList<>(transaction.locationsAdded.size());
-        final Collection<ImmutableTableLocationKey> removedImmutableKeys =
+        final Collection<ImmutableTableLocationKey> removedKeys =
                 new ArrayList<>(transaction.locationsRemoved.size());
 
-        // Process the accumulated adds and removes under a lock on `tableLocations` to keep modifications atomic to
-        // other holders of this lock.
-        synchronized (tableLocations) {
-            if (transaction.locationsAdded != EMPTY_TABLE_LOCATION_KEYS
-                    || transaction.locationsRemoved != EMPTY_TABLE_LOCATION_KEYS) {
-                for (TableLocationKey locationKey : transaction.locationsAdded) {
-                    locationCreatedRecorder = false;
-                    final Object result = tableLocations.putIfAbsent(locationKey, this::observeInsert);
-                    visitLocationKey(locationKey);
-                    if (locationCreatedRecorder) {
-                        verifyPartitionKeys(locationKey);
-                        addedImmutableKeys.add(toKeyImmutable(result));
-                    }
-                }
+        // Return early if there are no changes to process.
+        if (transaction.locationsAdded == EMPTY_TABLE_LOCATION_KEYS
+                && transaction.locationsRemoved == EMPTY_TABLE_LOCATION_KEYS) {
+            return;
+        }
 
-                for (TableLocationKey locationKey : transaction.locationsRemoved) {
-                    final Object removedLocation = tableLocations.remove(locationKey);
-                    if (removedLocation != null) {
-                        removedImmutableKeys.add(toKeyImmutable(locationKey));
-                    }
-                }
+        for (ImmutableTableLocationKey locationKey : transaction.locationsRemoved) {
+            final Object removedLocation = tableLocations.remove(locationKey);
```

Comment:
In both places that you remove, if the `removedLocation` is an instance of `ReferenceCountedTableLocaton` (or `AbstractTableLocation` if we keep the current designations), you decrement its reference count.
Same for the single place you call `makeTableLocation`, w.r.t. incrementing.